src/lib.rs

```diff
@@ -94,14 +99,104 @@ impl GitOid {
         BufReader<R>: std::io::Read,
     {
         let digest = hash_algo.create_digest();
-        match GitOid::generate_git_oid_from_buffer(digest, content, expected_length) {
-            Ok(v) => Ok(GitOid {
+        let v = GitOid::generate_git_oid_from_buffer(digest, content, expected_length)?;
+        Ok(GitOid {
+            hash_algorithm: hash_algo,
+            len: v.0,
+            value: v.1,
+        })
+    }
+
+    /// generate a bunch of `GitOid`s from a bunch of async
+    /// readers for a given algorithm
+    pub async fn new_from_async_readers<R, I>(
+        hash_algo: HashAlgorithm,
+        content: I,
+    ) -> IOResult<Vector<GitOid>>
+    where
+        R: AsyncReadExt + std::marker::Unpin,
+        I: IntoIterator<Item = (R, usize)>,
+    {
+        let digest = hash_algo.create_digest();
+        let mut futures = Vec::new();
+
+        // go get the futures for each hash operation
+        for (reader, expected_length) in content {
+            futures.push(GitOid::generate_git_oid_from_async_buffer(
+                digest.clone(),
+                reader,
+                expected_length,
+            ));
+        }
+
+        // create the return vector
+        let mut ret = Vector::new();
+
+        // go through each future and await the response
+        // the cool thing is that this will block on any given future
+        // but other futures may become satisfied so the look effectively
+        // blocks on the longest-to-satisfy future
+        for v in futures {
+            let (len, bytes) = v.await?;
+
+            ret.push_back(GitOid {
                 hash_algorithm: hash_algo,
-                len: v.0,
-                value: v.1,
-            }),
-            Err(e) => Err(e),
+                len: len,
+                value: bytes,
+            });
         }
+
+        Ok(ret)
+    }
+
+    /// the async version of generating a git_oid from a buffer
+    async fn generate_git_oid_from_async_buffer<R>(
+        mut digest: Box<dyn DynDigest>,
+        mut reader: R,
+        expected_length: usize,
+    ) -> IOResult<(usize, [u8; NUM_HASH_BYTES])>
+    where
+        R: AsyncReadExt + std::marker::Unpin,
+    {
+        let prefix = format!("blob {}\0", expected_length);
+
+        let mut buf = [0u8; 4096]; // Linux default page size is 4096
+        let mut amount_read: usize = 0;
+
+        // set the prefix
+        digest.update(prefix.as_bytes());
+
+        // keep reading the input until there is no more
+        loop {
+            match reader.read(&mut buf).await? {
+                // done
+                0 => break,
+
+                // update the hash and accumulate the count
+                size => {
+                    digest.update(&buf[..size]);
+                    amount_read = amount_read + size;
+                }
+            }
+        }
```

Comment:
When I looked at the async `BufReader`, I could not find the `read` function... also given the size of the reads (4K) and the likelihood that the reads will be coming from the filesystem, I'm not sure where buffering would buy us anything. Happy to be educated.

Reply:
The `BufReader` type wraps a type implementing `AsyncRead` to provide buffering for it. By default it is [8 kilobytes](https://doc.rust-lang.org/src/std/sys_common/io.rs.html#3) unless we're on specific architectures.

While it implements `AsyncRead` (you can find `async fn read` implemented on `AsyncReadExt`), it also exposes "buffered read" APIs through the `AsyncBufRead{Ext,}` traits. This allows performing buffered reads in two steps:

1. Call `fill_buf` to fill the buffer with as many bytes as the underlying reader will provide you. This gives you back a reference to a filled buffer.
2. Once you have the filled buffer you can use those bytes. Once you're done with them you should call `consume` to indicate how many bytes you've used from the buffer, so that a next call to `fill_buf` knows how much space to reuse or reallocate on the buffer.

Put together that provides a system quite similar to what you've manually authored here, but using a more standard construct.